### DISCUSSION TITLE
Database Providers

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3554] [General]
In the transition to .NET 8 (and .NET 7) there was a timing/dependency issue identified with the installation of Database Provider packages which caused upgrade issues for the framework. The solution which was implemented was to install all Database Providers very early in the startup logic so that it is guaranteed that all required dependencies are available to the framework. This successfully resolved the framework upgrade problem.

The side effect of this solution is that the Database Provider Nuget packages still needed to be included in the wwwroot/Packages temporary location so they could be copied to the private /Packages folder and installed (and then removed). This resulted in the Database Provider Nuget packages always appearing as modified files in Git - and there is no way to untrack them globally, so care had to be taken to not accidentally remove them from the source repo when submitting PRs. 

#3553 resolves this issue by eliminating the need for Database Provider Nuget packages. Since the new approach installs all of the database providers by default, it is possible to simply deploy the assemblies to the Oqtane /bin directly when the Oqtane.Databases.sln is built. This eliminates the need for the temporary wwwroot/Packages folder, the copying of the Nuget packages to the /Packages folder, and the issues with GitHub source control tracking these files.

However, the one side effect is that if a developer clones the source code from GitHub for a new installation, opens the Oqtane.sln and hits F5, they will encounter an error because the Database Provider assemblies are not in the /bin folder. To resolve this they would need to open Oqtane.Databases.sln and build that solution first in order to create the Database Provider assemblies. 

This could be resolved by including the Oqtane.Database.* project files in the Oqtane.sln - however the reason why a separate solution was created for the database projects was to streamline the build time by avoiding building projects which rarely change, and avoiding the display of compiler warnings in the database providers which are related to the utilization of internal Entity Framework Core APIs. So my preference is to not add the Oqtane.Database.* project files to the Oqtane.sln - but this means that the clone and run experience will not be as seamless as it was previously. 

This seems to be one of those problems where every solution has side effects. So it comes down to prioritizing which use cases are the most important. Ensuring a seamless framework upgrade experience is definitely the #1 priority. I would like to hear feedback on the other items I described above to determine if #3553 should be rolled back in favor of the previous behavior, or if the new behavior is acceptable.

**Comment by sbwalker** [comment]
#3555 suppresses the compiler warnings in the Database Provider projects... so the only issue with including the Oqtane.Database.* project files in the Oqtane.sln is the extra build time when doing a Rebuild (Build will only compile projects which have actually changed). So perhaps its fine to add them to the Oqtane.sln. Please share your preferences...

**Reply by thabaum** [reply]
I am in agreement that with the Database project being included in the Oqtane project there will be less out of the box developer issues for getting the project fired up immediately upon downloading the source repo off GitHub.  I think it is a small enough project it should not add too much weight in build time.

I see the value in having it as an independent project to lower the time of the main project build.  I do like how Oqtane has quick builds.

I have included the Oqtane.Maui project into Oqtane solution in the past and worked with an Oqtane Blazor Server Maui Hybrid approach using two instances of visual studio with the Oqtane solution and editing the Maui project to work with the main solution to create and run a Maui and Server/Client project.   

The Oqtane solution did take a little more time to build so I figured this was why it is not included.  Plus not everyone will use Oqtane for store apps or desktop apps.  I almost think the Oqtane.Maui project should be included as well to help showcase the mobile native app of Maui Blazor Hybrid application development.

Bottom line keeping it simple for developers to understand out of the box holds some real value.

I am OK either way.

Cheers!

**Reply by leigh-pointer** [reply]
@sbwalker personally leave as is and introduce a new sln file! If the dbs are added to the current solution they need to be set to not compile in the debug and release builds.

**Reply by sbwalker** [reply]
@leigh-pointer I assume you mean introduce a new sln file which contains the database projects... and developers who are new to Oqtane would need to know that they should use this sln file (rather than Oqtane.sln) to get the initial installation up and running. I am not sure if this is a great option as it still makes "clone and run" challenging for new developers.

If the database providers are added to the Oqtane.sln then the only scenario where the dev experience will be slower is when doing a full Build / Rebuild Solution in Visual Studio. If you do a basic Build / Build Solution or just hit F5, the database providers will not be compiled (because the code has not changed) - so it will not affect developer productivity. Also if your module or theme project is linked to Oqtane.Server it should not affect developer productivity either.

I will add them to the Oqtane.sln so that it is possible to evaluate the actual behavior rather than speculating.
